### DISCUSSION TITLE
Nostr Relay Service for Fedimints

### DISCUSSION THREAD

**Discussion by okjodom** [fedimint/fedimint#1604] [Fedimint Modules]
Opening this discussion thread to consider use cases, need and possibility of Nostr relays as a way to move generic data across the Fedimint boundary

**Comment by BitcoinBrief** [comment]
p2p marketplace.  Nostr relay module to post trade offers and facilitate trader chat.  Combine with escrow module (https://github.com/fedimint/fedimint/discussions/1513#discussion-4789028) to secure trade amounts and deposits.  Guardians would need to remain anon and/or in large numbers to protect against FinCEN, regulatory challenges.

**Comment by elsirion** [comment]
There are two distinct use cases that are interesting:

1. **Being a relay**: A federation acting as a relay could guarantee censorship resistance and ordering of events. While this *might* be interesting for some applications it comes at a hefty price since the throughput of the consensus protocol is rather limited (ok for financial tx, possibly too much for a social network).
As a middle ground guardians of a federation could run relays. ~that sync events between each other~ and users connect to all of them. This would retain the censorship resistance while trading strict ordering of events for efficiency.

2. **Federations as event sources/recipients**: Since Nostr events are signed with Schnorr signatures federations can use FROST/ROAST to generate their own Nostr events whose authenticity is ensured by the federation. This might be useful as trusted price oracles, order books, etc. One could even imagine Nostr being used as a chat-based user interface for applications running on Fedimint, akin to LNTxBot. I think that's the direction https://github.com/fedimint/fedimint/discussions/1604#discussioncomment-4925409 also goes into.

**Reply by dpc** [reply]
> Being a relay: A federation acting as a relay could guarantee censorship resistance 

Isn't that equal to just having users publish to N relays ran by the N people? In nostr just one relay needs to be honest and relay a given message, which is much easier than typical federation T of M.

**Reply by elsirion** [reply]
Right, I got confused.

**Comment by dpc** [comment]
I'm not against nostr by any means (BTW, follow me at npub1yarprdrj3vxyp0su7xqd3u6z047na6792ezas6dqqt5tv4lce4ssjt0gd0 :D) , but I don't see any synergy between them. They can work along-side just fine, but they aim at a completely different things in different realms. I even sense a deep impedance mismatch between them, though I can't tell exactly why just yet - just an intuition.

At it's core nostr is just self-signed documents, and servers for relaying them. It's intentionally very, very simple, and it's more about community around it and ease of building on top of it then anything technical inside it. Fedimint is very different where it's complex crypto, and consensus protocols integrated into modular but very tight whole.

In Fedimint everything is rooted in a trust in T of M guardians. In nostr everything is about slapping your message at N relays so that anyone can fetch it if they are interested and it's scalable and censorship resistant by redundancy. There's almost no need for any trust.

It seems to me (I might be wrong, as always) that whatever would require nostr's scalable distribution abilities can be done by the interested (and incentivized) Fedimint client/users themselves, so that Fedimint does not have even know  about it. You want to publish your Fedimint smart contract offer - publish it on nostr, sure, but under your own id, with your own client, over your own relays, etc. (which can be integrated in a client/product supporting both at the same time).

**Comment by jkitman** [comment]
Fedimint consensus couldn't handle all events being put into consensus but here are some possible integration points:

- Spam prevention.  I believe this will be the biggest issue for NOSTR which is currently being solved by paying LN to sign up to some relays.  You could potentially replace LN with ecash, perhaps even paid on each post or after a certain number of posts.
- Availability.  While it's impossible to store all events in consensus, if you could order them according to some scheme and ensure that all Fedimint relays store them in "blocks" you could challenge or pay relays periodically to ensure they are still storing events.
- Privacy.  NOSTR offers no network-level privacy and Tor has timing analysis issues.  However, there are many federated schemes based on Chaum's DCnets that can offer perfect privacy within a given time period which may work well for Twitter-like data with schemes described in https://people.csail.mit.edu/devadas/pubs/atom.pdf.

That said, I would agree that there are challenges with all of the above and it's probably too early in NOSTR + Fedimint history to see where things will end up.

**Reply by dpc** [reply]
> Spam prevention. I believe this will be the biggest issue for NOSTR which is currently being solved by paying LN to sign up to some relays. You could potentially replace LN with ecash, perhaps even paid on each post or after a certain number of posts.

Makes sense, but the relay can be Fedimint client, and payer would be Fedimint client, so it can be external project.

Also - if the payment was to be per-post ... I'm quite sure a single relay scales waaay better than a federation, and typical nostr post has barely more data than a single used nonce, so until we have some key rotation to prune history, I don't think this will work due to amount of data required.


Once we have mint module key rotation we could have two mint module instances: slow and fast. Where fast would eg. do key rotation weekly, and was for automated systems that would re-mint their outstanding notes daily. This way federation could handle lots of volume.

> Availability. While it's impossible to store all events in consensus, if you could order them according to some scheme and ensure that all Fedimint relays store them in "blocks" you could challenge or pay relays periodically to ensure they are still storing events.

Hmmm... We could eventually have a extendable secondary storage architecture, and nostr could be one of the implementations.

>  However, there are many federated schemes based on Chaum's DCnets that can offer perfect privacy within a given time period which may work well for Twitter-like data with schemes described

OK. That sounds like something that might be a good use-case for direct Fedimint+nostr integration.

**Comment by jkitman** [comment]
Since Nostr npubs are Schnorr, you could have the federation threshold-sign to post messages from a shared Nostr account.

Could be an interesting way for a fed to communicate with the outside world.

**Comment by EthnTuttle** [comment]
If anyone comes across this thread or was active here, I am planning to work on an implementation of nsotr+fedimint over the next several months. Please reach out if you have more ideas or would like to help.
